### PR TITLE
Feature/0.14.0

### LINKED ISSUE #27: Add option to use `npm set` and `npm adduser`
It sounds reasonable to be able to use `npm set` on build servers. I'd like for instance to be able to disable the progress just before running npm install as that would improve the speed significantly. Having an extra cake command to use "npm set", would enable us to configure the npm settings on the build server. 

http://biercoff.com/how-to-crazy-easily-speed-up-your-npm-install-speed/

This would also allow for setting the registry to be used on the build server/agent for npm to use. Of course you would need a login command there as well ;)

### LINKED ISSUE #28: add support for npm update
There are currently aliases for running Npm scripts and running Npm install. There is no alias or available way to call Npm update. https://docs.npmjs.com/cli/update 

In CI builds I would like to run Npm update from Cake so that I don't have to clear out all node_modules on each build and then install to ensure I have the latest versions. I should be able to run Npm update to update out of date packages without re-installing everything.

### LINKED ISSUE #68: Support npm version
It would be helpful to support `npm version` to be able to bump the version easily from Cake.
Reference: https://docs.npmjs.com/cli/version

### LINKED ISSUE #75: Add the ability to specify --production in NpmCiSettings
Add the ability to specify `--production` in `NpmCiSettings`

### LINKED ISSUE #79: Build against Cake 0.28.0
Build against Cake 0.28.0

### LINKED ISSUE #80: Target .NET Standard 2.0
Target .NET Standard 2.0

### PR DESCRIPTION
**Pull request by Philo**
Fixes #68 
Fixes #27 
Fixes #75 
Fixes #28 
Fixes #79
Fixes #80

### CODE REVIEW

**Comment by pascalberger** [comment]
GitReleaseManager (the part responsible for release notes creation in Cake.Recipe), just looks for a milestone with the same version as the current release branch. Then it checks the issues with labels as defined [here](https://github.com/cake-contrib/Cake.Npm/blob/develop/GitReleaseManager.yaml). Note that an issue can only have a single label. Based on this it creates the GitHub draft release including release notes. You can use the `Fixes #123` notation in pull requests, to close issues as soon as a pull request is merged.

Cake.Recipe release process is documented [here](https://cake-contrib.github.io/Cake.Recipe/docs/usage/creating-release).

The point with multiple issues solved in one PR is that it makes review not easier :) But will have a look at it later tonight.

**Comment by Philo** [comment]
Ah yes, I remember seeing that somewhere and thinking wth - not my username and password for github into an ENV var - no thanks, even temporarily.  Plus I also use 2FA on my account and I dont know if that is supported.

**Comment by pascalberger** [comment]
You can, and in case of 2FA need to, use an [access token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/) for the password, which you can delete immediately after use. But I can create the release afterwards if you prefer.

**Comment by Philo** [comment]
@gep13 - part paranoid, part the sheer amount of GDPR and compliance to data breach and governance work I've been involved in recently.  If sensitive information is ever available system-wide, even for a short period of time I can effectively no longer trust that the information is safe, secured and has not been breached, the new regulations require that companies give notice of potential breaches if personal data or passwords were left unsecured at all, for any period of time.
  
I treat my important online accounts in the same way I would treat my bank account, my passwords are highly complex such that I dont even know what they are and rely on a yubikey static password feature to part fill my passwords.

I realise this is pretty extreme, but its something I wont want to compromise.  Assuming that instead of a username and password this could be done using a short lived PAT (personal access token) then I would be happy to go through the extra steps of creating a short lived PAT for use in this scenario.  Perhaps this is possible:

image

But what scopes would need to be granted (temporarily) to support the required functionality of GitReleaseManager?

**Comment by gep13** [comment]
It is absolutely possible to use a PAT. The parameters to GRM say username and password, but you can pass a PAT and it will happily work. That is actually what I do.

In terms of what scopes are required, I would need to get back to you on that.

There would always be the option of running GRM directly from the command line, rather than through Cake, and then you wouldn’t be storing the PAT anywhere, but passing the parameters at the command line.

**Comment by pascalberger** [comment]
I see we need GDPR compatibility in Cake.Recipe 😉 

But should be fairly easy to also take build parameters into account [here](https://github.com/cake-contrib/Cake.Recipe/blob/develop/Cake.Recipe/Content/credentials.cake#L155-L156)

File: src/Cake.Npm/NpmUpdateAliases.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+            
+            context.NpmUpdate(new NpmUpdateSettings());
+        }
+
+        /// <summary>
+        /// Updates packages using the speUpdatefied settings.
```
Comment (pascalberger) [inline src/Cake.Npm/NpmUpdateAliases.cs]: Typo `speUpdatefied`

File: src/Cake.Npm/NpmUpdateAliases.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="settings">The settings.</param>
+        /// <example>
+        /// <para>Use speUpdatefic log level ('npm Update')</para>
```
Comment (pascalberger) [inline src/Cake.Npm/NpmUpdateAliases.cs]: I don't understand this paragraph

File: src/Cake.Npm/NpmUpdateAliases.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        /// <code>
+        /// <![CDATA[
+        ///     var settings = new NpmUpdateSettings();
```
Comment (pascalberger) [inline src/Cake.Npm/NpmUpdateAliases.cs]: Would be good to do something with the settings (e.g. set `global` flag)

File: src/Cake.Npm.Tests/Update/NpmUpdateToolTests.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
+﻿using Cake.Npm.Update;
+using Xunit;
+
+namespace Cake.Npm.Tests.Update
+{
+    public class NpmUpdateToolTests
+    {
+        public sealed class TheAutomatedInstallMethod
```
Comment (pascalberger) [inline src/Cake.Npm.Tests/Update/NpmUpdateToolTests.cs]: Copy / Paste error? Or to which `AutomatedInstall` method does this reference

File: src/Cake.Npm/Ci/NpmCiSettingsExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+﻿using Cake.Npm.Ci;
+
+namespace Cake.Npm.Install
```
Comment (pascalberger) [inline src/Cake.Npm/Ci/NpmCiSettingsExtensions.cs]: Should be in Cake.Npm.Ci namespace

File: src/Cake.Npm.Tests/Ci/NpmCiToolTests.cs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using Cake.Core.Diagnostics;
+using Cake.Npm.Install;
```
Comment (pascalberger) [inline src/Cake.Npm.Tests/Ci/NpmCiToolTests.cs]: Should not be required if settings are in `Cake.Npm.Ci` namepsace

File: src/Cake.Npm.Tests/Set/NpmSetToolTests.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
+﻿using Cake.Npm.Set;
+using Shouldly;
+using Xunit;
+
+namespace Cake.Npm.Tests.Set
+{
+    public class NpmSetToolTests
+    {
+        public sealed class TheAutomatedInstallMethod
```
Comment (pascalberger) [inline src/Cake.Npm.Tests/Set/NpmSetToolTests.cs]: Copy / Paste error? Or to which `AutomatedInstall` method does this reference

File: src/Cake.Npm/NpmSetAliases.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                Global = global
+            });
+        }
+
+        /// <summary>
+        /// Sets an npm configuration setting.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="settings">The settings.</param>
+        /// <param name="key">The key</param>
```
Comment (pascalberger) [inline src/Cake.Npm/NpmSetAliases.cs]: Remove as this parameter doesn't exist

File: src/Cake.Npm/NpmSetAliases.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        /// <param name="value">The value</param>
```
Comment (pascalberger) [inline src/Cake.Npm/NpmSetAliases.cs]: Remove as this parameter doesn't exist

File: src/Cake.Npm/NpmSetAliases.cs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        /// <param name="globally">Set globally</param>
```
Comment (pascalberger) [inline src/Cake.Npm/NpmSetAliases.cs]: Remove as this parameter doesn't exist

File: src/Cake.Npm/NpmSetAliases.cs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        /// <example>
+        /// <para>Use speSetfic log level ('npm Set')</para>
```
Comment (pascalberger) [inline src/Cake.Npm/NpmSetAliases.cs]: I don't understand this sentence

File: src/Cake.Npm/NpmSetAliases.cs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        /// <code>
+        /// <![CDATA[
+        ///     var settings = new NpmSetSettings();
```
Comment (pascalberger) [inline src/Cake.Npm/NpmSetAliases.cs]: Should set values on the settings object.

File: src/Cake.Npm/NpmSetAliases.cs [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        ///     NpmSet(settings, "progress", "false", true);
```
Comment (pascalberger) [inline src/Cake.Npm/NpmSetAliases.cs]: This call is not valid

File: src/Cake.Npm.Tests/Version/NpmVersionToolTests.cs [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+                    "    node: '8.11.1',",
+                    "    openssl: '1.0.2o',",
+                    "    tz: '2017c',",
+                    "    unicode: '10.0',",
+                    "    uv: '1.19.1',",
+                    "    v8: '6.2.414.50',",
+                    "    zlib: '1.2.11'",
+                    "}",
+"",
+"",
```
Comment (pascalberger) [inline src/Cake.Npm.Tests/Version/NpmVersionToolTests.cs]: Is the indentation here intentional?

File: src/Cake.Npm/NpmVersionAliases.cs [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+            
+            return context.NpmVersion(new NpmVersionSettings());
+        }
+
+        /// <summary>
+        /// Versions packages using the speVersionfied settings.
```
Comment (pascalberger) [inline src/Cake.Npm/NpmVersionAliases.cs]: Typo `speVersionfied`

File: src/Cake.Npm/NpmVersionAliases.cs [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="settings">The settings.</param>
+        /// <example>
+        /// <para>Use speVersionfic log level ('npm Version')</para>
```
Comment (pascalberger) [inline src/Cake.Npm/NpmVersionAliases.cs]: I don't understand this sentence

File: src/Cake.Npm/NpmVersionAliases.cs [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        /// <code>
+        /// <![CDATA[
+        ///     var settings = new NpmVersionSettings();
```
Comment (pascalberger) [inline src/Cake.Npm/NpmVersionAliases.cs]: Should set a property on the settings instance

File: VersionAssemblyInfo.cs [hunk 18]
```diff
@@ -0,0 +1,11 @@
+//------------------------------------------------------------------------------
```
Comment (pascalberger) [inline VersionAssemblyInfo.cs]: I don' think this file should be under version control

File: src/Cake.Npm.Tests/Version/NpmVersionToolTests.cs (hunk 14 above)
Comment (Philo) [reply src/Cake.Npm.Tests/Version/NpmVersionToolTests.cs]: The indentation is intentional, to duplicate the slightly daft output from the `npm version` command.  Annoyingly, it returns a json fragment, unless you have an older version of npm and then it appends a block of text about updating that means we cant just treat the process output as json.

File: src/Cake.Npm/NpmUpdateAliases.cs (hunk 1 above)
Comment (Philo) [reply src/Cake.Npm/NpmUpdateAliases.cs]: corrected typo

File: src/Cake.Npm/NpmUpdateAliases.cs (hunk 2 above)
Comment (Philo) [reply src/Cake.Npm/NpmUpdateAliases.cs]: removed

File: src/Cake.Npm/NpmUpdateAliases.cs (hunk 3 above)
Comment (Philo) [reply src/Cake.Npm/NpmUpdateAliases.cs]: added example

File: src/Cake.Npm.Tests/Update/NpmUpdateToolTests.cs (hunk 4 above)
Comment (Philo) [reply src/Cake.Npm.Tests/Update/NpmUpdateToolTests.cs]: yea, updated all the instances where this occurs

File: src/Cake.Npm/Ci/NpmCiSettingsExtensions.cs (hunk 5 above)
Comment (Philo) [reply src/Cake.Npm/Ci/NpmCiSettingsExtensions.cs]: moved

File: src/Cake.Npm.Tests/Ci/NpmCiToolTests.cs (hunk 6 above)
Comment (Philo) [reply src/Cake.Npm.Tests/Ci/NpmCiToolTests.cs]: idd

File: src/Cake.Npm/NpmVersionAliases.cs (hunk 15 above)
Comment (Philo) [reply src/Cake.Npm/NpmVersionAliases.cs]: corrected

File: src/Cake.Npm/NpmVersionAliases.cs (hunk 16 above)
Comment (Philo) [reply src/Cake.Npm/NpmVersionAliases.cs]: removed

File: src/Cake.Npm/NpmVersionAliases.cs (hunk 17 above)
Comment (Philo) [reply src/Cake.Npm/NpmVersionAliases.cs]: removed the alias that uses the settings has there aren't any settings properties to set for `npm version`

**Comment by Philo** [comment]
@pascalberger - went through and fixed typos and resolved poor documentation and comment issues

File: src/Cake.Npm/Namespaces.cs [hunk 19]
```diff
@@ -59,3 +59,51 @@ internal class NamespaceDoc
     {
     }
 }
+
+// ReSharper disable once CheckNamespace
+namespace Cake.Npm.Ci
+{
+    /// <summary>
+    /// This namespace contain types used for running npm scripts.
```
Comment (pascalberger) [inline src/Cake.Npm/Namespaces.cs]: Wrong comment

File: src/Cake.Npm/Namespaces.cs [hunk 20]
```diff
@@ -59,3 +59,51 @@ internal class NamespaceDoc
...
+    internal class NamespaceDoc
+    {
+    }
+}
+
+// ReSharper disable once CheckNamespace
+namespace Cake.Npm.Set
+{
+    /// <summary>
+    /// This namespace contain types used for running npm scripts.
```
Comment (pascalberger) [inline src/Cake.Npm/Namespaces.cs]: Wrong comment

File: src/Cake.Npm/Namespaces.cs [hunk 21]
```diff
@@ -59,3 +59,51 @@ internal class NamespaceDoc
...
+    internal class NamespaceDoc
+    {
+    }
+}
+
+// ReSharper disable once CheckNamespace
+namespace Cake.Npm.Version
+{
+    /// <summary>
+    /// This namespace contain types used for running npm scripts.
```
Comment (pascalberger) [inline src/Cake.Npm/Namespaces.cs]: Wrong comment

File: src/Cake.Npm/Namespaces.cs [hunk 22]
```diff
@@ -59,3 +59,51 @@ internal class NamespaceDoc
...
+    internal class NamespaceDoc
+    {
+    }
+}
+
+// ReSharper disable once CheckNamespace
+namespace Cake.Npm.Update
+{
+    /// <summary>
+    /// This namespace contain types used for running npm scripts.
```
Comment (pascalberger) [inline src/Cake.Npm/Namespaces.cs]: Wrong comment

File: src/Cake.Npm/NpmSetAliases.cs (hunk 12 above)
Comment (pascalberger) [reply src/Cake.Npm/NpmSetAliases.cs]: This is still an issue. Should be something like:

```
var settings = new NpmSetSettings();
settings.Key = "progress";
settings.Value = "false";
NpmSet(settings);
```

File: src/Cake.Npm/NpmSetAliases.cs (hunk 13 above)
Comment (pascalberger) [reply src/Cake.Npm/NpmSetAliases.cs]: This is still an issue (wrong example for this alias)

File: src/Cake.Npm/Set/NpmSetSettings.cs [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+        /// The config value
+        /// </summary>
+        public string Value { get; set; }
+
+        /// <summary>
+        /// Determines whether to set the config key value globally
+        /// </summary>
+        public bool Global { get; set; }
+
+        /// <inheritdoc cref="NpmSettings.EvaluateCore" />
```
Comment (pascalberger) [inline src/Cake.Npm/Set/NpmSetSettings.cs]: cref should not be required

File: src/Cake.Npm/Update/NpmUpdateSettings.cs [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+
+        /// <summary>
+        /// Gets a value indicating whether to update globally installed packages.
+        /// </summary>
+        public bool Global { get; internal set; }
+
+        /// <summary>
+        /// Evaluates the settings and writes them to <paramref name="args"/>.
+        /// </summary>
+        /// <param name="args">The argument builder into which the settings should be written.</param>
```
Comment (pascalberger) [inline src/Cake.Npm/Update/NpmUpdateSettings.cs]: You can unse `<inheritdoc />` here

File: src/Cake.Npm/Ci/NpmCiSettings.cs [hunk 25]
```diff
@@ -12,5 +15,24 @@ public NpmCiSettings()
...
+
+        /// <summary>
+        /// Gets a value indicating whether npm should ignore modules listed in <code>devDependencies</code>.
+        /// </summary>
+        public bool Production { get; set; }
+
+        /// <summary>
+        /// Evaluates the settings and writes them to <paramref name="args"/>.
+        /// </summary>
+        /// <param name="args">The argument builder into which the settings should be written.</param>
```
Comment (pascalberger) [inline src/Cake.Npm/Ci/NpmCiSettings.cs]: You can unse `<inheritdoc />` here

**Comment by pascalberger** [comment]
@Philo LGTM! Do you want me to do the release?

**Comment by Philo** [comment]
I was hoping you would offer to do the release!! - yes please. :)

**Pull request merged**